src/stamp/encoding/encoder/__init__.py

```diff
@@ -0,0 +1,193 @@
+import os
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import TypeVar
+
+import h5py
+import numpy as np
+import pandas as pd
+import torch
+from torch import Tensor, nn
+from torch._prims_common import DeviceLikeType  # type: ignore
+from tqdm import tqdm
+
+import stamp
+from stamp.cache import get_processing_code_hash
+from stamp.encoding.config import EncoderName
+from stamp.modeling.data import CoordsInfo, PandasLabel, get_coords
+from stamp.preprocessing.config import ExtractorName
+
+EncoderModel = TypeVar("EncoderModel", bound=nn.Module)
+
+__author__ = "Juan Pablo Ricapito"
+__copyright__ = "Copyright (C) 2025 Juan Pablo Ricapito"
+__license__ = "MIT"
+
+
+class Encoder(ABC):
+    def __init__(
+        self,
+        model,
+        identifier: EncoderName,
+        precision: torch.dtype,
+        required_extractor: list[ExtractorName],
+    ):
+        self.model = model
+        self.identifier = identifier
+        self.precision = precision
+        self.required_extractor = required_extractor
+
+    def encode_slides(
+        self,
+        output_dir: Path,
+        feat_dir: Path,
+        device: DeviceLikeType,
+        generate_hash: bool,
+        **kwargs,
+    ) -> None:
+        """General method to encode slides."""
+        output_file = self._generate_output_path(
+            output_dir=output_dir, generate_hash=generate_hash
+        )
+
+        slide_dict = {}
+        self.model.to(device).eval()
+
+        if self.precision == torch.float16:
+            self.model.half()
+
+        if os.path.exists(output_file):
+            tqdm.write(f"Output file {output_file} already exists, skipping")
+            return
+
+        for tile_feats_filename in tqdm(os.listdir(feat_dir), desc="Processing slides"):
+            h5_path = os.path.join(feat_dir, tile_feats_filename)
+            slide_name: str = Path(tile_feats_filename).stem
+
+            try:
+                feats, coords = self._validate_and_read_features(h5_path)
+            except ValueError as e:
+                tqdm.write(s=str(e))
+                continue
+
+            slide_embedding = self._generate_slide_embedding(
+                feats, device, coords=coords
+            )
+            slide_dict[slide_name] = {"feats": slide_embedding}
+
+        self._save_features(output_file, entry_dict=slide_dict)
+
+    def encode_patients(
+        self,
+        output_dir: Path,
+        feat_dir: Path,
+        slide_table_path: Path,
+        patient_label: PandasLabel,
+        filename_label: PandasLabel,
+        device: DeviceLikeType,
+        generate_hash: bool,
+        **kwargs,
+    ) -> None:
+        """General method to encode patients with features only.
+        Override this method if coords are needed for encoding"""
+        output_file = self._generate_output_path(
+            output_dir=output_dir, generate_hash=generate_hash
+        )
+
+        patient_dict = {}
```

Comment:
Please provide type annotation (and optionally a comment explaining the dict, e.g. `# maps patient ID -> patient embedding`)